README.md
# hadoop cookbook

[![Cookbook Version](http://img.shields.io/cookbook/v/hadoop.svg)](https://supermarket.chef.io/cookbooks/hadoop)
[![Apache License 2.0](http://img.shields.io/badge/license-apache%202.0-green.svg)](http://opensource.org/licenses/Apache-2.0)
[![Build Status](http://img.shields.io/travis/caskdata/hadoop_cookbook.svg)](http://travis-ci.org/caskdata/hadoop_cookbook)
[![Code Climate](https://codeclimate.com/github/caskdata/hadoop_cookbook/badges/gpa.svg)](https://codeclimate.com/github/caskdata/hadoop_cookbook)

## Requirements

This cookbook may work on earlier versions, but these are the minimal tested versions.

* Chef 11.4.0+
* CentOS 6.4+
* Debian 6.0+
* Ubuntu 12.04+

This cookbook assumes that you have a working Java installation. It has been tested using version `1.21.2` of the `java` cookbook, using Oracle JDK 7. If you plan on using Hive with a database other than the embedded Derby, you will need to provide it and set it up prior to starting Hive Metastore service.

## Usage

This cookbook is designed to be used with a wrapper cookbook or a role with settings for configuring Hadoop. The services should work out of the box on a single host, but little validation is done that you have made a working Hadoop configuration. The cookbook is attribute-driven and is suitable for use via either `chef-client` or `chef-solo` since it does not use any server-based functionality. The cookbook defines service definitions for each Hadoop service, but it does not enable or start them, by default.

For more information, read the [Wrapping this cookbook](https://github.com/caskdata/hadoop_cookbook/wiki/Wrapping-this-cookbook) wiki entry.

## Attributes

Attributes for this cookbook define the configuration files for Hadoop and its various services. Hadoop configuration files are XML files, with name/value property pairs. The attribute name determines which file the property is placed and the property name. The attribute value is the property value. The attribute `hadoop['core_site']['fs.defaultFS']` will configure a property named `fs.defaultFS` in `core-site.xml` in `hadoop['conf_dir']`. All attribute values are taken as-is and only minimal configuration checking is done on values. It is up to the user to provide a valid configuration for your cluster.

Attribute Tree | File | Location
-------------- | ---- | --------
flume['flume_conf'] | flume.conf | `flume['conf_dir']`
hadoop['capacity_scheduler'] | capacity-scheduler.xml | `hadoop['conf_dir']`
hadoop['container_executor'] | container-executor.cfg | `hadoop['conf_dir']`
hadoop['core_site'] | core-site.xml | `hadoop['conf_dir']`
hadoop['fair_scheduler'] | fair-scheduler.xml | `hadoop['conf_dir']`
hadoop['hadoop_env'] | hadoop-env.sh | `hadoop['conf_dir']`
hadoop['hadoop_metrics'] | hadoop-metrics.properties | `hadoop['conf_dir']`
hadoop['hadoop_policy'] | hadoop-policy.xml | `hadoop['conf_dir']`
hadoop['hdfs_site'] | hdfs-site.xml | `hadoop['conf_dir']`
hadoop['log4j'] | log4j.properties | `hadoop['conf_dir']`
hadoop['mapred_env'] | mapred-env.sh | `hadoop['conf_dir']`
hadoop['mapred_site'] | mapred-site.xml | `hadoop['conf_dir']`
hadoop['yarn_env'] | yarn-env.sh | `hadoop['conf_dir']`
hadoop['yarn_site'] | yarn-site.xml | `hadoop['conf_dir']`
hbase['hadoop_metrics'] | hadoop-metrics.properties | `hbase['conf_dir']`
hbase['hbase_env'] | hbase-env.sh | `hbase['conf_dir']`
hbase['hbase_policy'] | hbase-policy.xml | `hbase['conf_dir']`
hbase['hbase_site'] | hbase-site.xml | `hbase['conf_dir']`
hbase['jaas'] | jaas.conf | `hbase['conf_dir']`
hbase['log4j'] | log4j.properties | `hbase['conf_dir']`
hive['hive_env'] | hive-env.sh | `hive['conf_dir']`
hive['hive_site'] | hive-site.xml | `hive['conf_dir']`
hive['jaas'] | jaas.conf | `hive['conf_dir']`
hive2['hive_env'] | hive-env.sh | `hive2['conf_dir']`
hive2['hive_site'] | hive-site.xml | `hive2['conf_dir']`
hive2['jaas'] | jaas.conf | `hive2['conf_dir']`
oozie['oozie_env'] | oozie-env.sh | `oozie['conf_dir']`
oozie['oozie_site'] | oozie-site.xml | `oozie['conf_dir']`
spark['log4j'] | log4j.properties | `spark['conf_dir']`
spark['metrics'] | metrics.properties | `spark['conf_dir']`
spark['spark_env'] | spark-env.sh | `spark['conf_dir']`
storm['storm_env'] | storm-env.sh | `storm['conf_dir']`
storm['storm_env'] | storm_env.ini | `storm['conf_dir']`
storm['storm_conf'] | storm.yaml | `storm['conf_dir']`
tez['tez_env'] | tez-env.sh | `tez['conf_dir']`
tez['tez_site'] | tez-site.xml | `tez['conf_dir']`
zookeeper['jaas'] | jaas.conf | `zookeeper['conf_dir']`
zookeeper['log4j'] | log4j.properties | `zookeeper['conf_dir']`
zookeeper['zoocfg'] | zoo.cfg | `zookeeper['conf_dir']`

### Distribution Attributes

* `hadoop['distribution']` - Specifies which Hadoop distribution to use, currently supported: cdh, hdp, bigtop. Default `hdp`
* `hadoop['distribution_version']` - Specifies which version of `hadoop['distribution']` to use. Default `2.0` if `hadoop['distribution']` is `hdp`, `5` if `hadoop['distribution']` is `cdh`, and `0.8.0` if `hadoop['distribution']` is `bigtop`.  It can also be set to `develop` when `hadoop['distribution']` is `bigtop` to allow installing from development repos without gpg validation.

#### APT-specific settings

* `hadoop['apt_repo_url']` - Provide an alternate apt installation source location. If you change this attribute, you are expected to provide a path to a working repo for the `hadoop['distribution']` used. Default: `nil`
* `hadoop['apt_repo_key_url']` - Provide an alternative apt repository key source location. Default `nil`

#### RPM-specific settings

* `hadoop['yum_repo_url']` - Provide an alternate yum installation source location. If you change this attribute, you are expected to provide a path to a working repo for the `hadoop['distribution']` used. Default: `nil`
* `hadoop['yum_repo_key_url']` - Provide an alternative yum repository key source location. Default `nil`

### Global Configuration Attributes

* `hadoop['conf_dir']` - The directory used inside `/etc/hadoop` and used via the alternatives system. Default `conf.chef`
* `hbase['conf_dir']` - The directory used inside `/etc/hbase` and used via the alternatives system. Default `conf.chef`
* `hive['conf_dir']` - The directory used inside `/etc/hive` and used via the alternatives system. Default `conf.chef`
* `oozie['conf_dir']` - The directory used inside `/etc/oozie` and used via the alternatives system. Default `conf.chef`
* `tez['conf_dir']` - The directory used inside `/etc/tez` and used via the alternatives system. Default `conf.chef`
* `spark['conf_dir']` - The directory used inside `/etc/spark` and used via the alternatives system. Default `conf.chef`
* `storm['conf_dir']` - The directory used inside `/etc/storm` and used via the alternatives system. Default `conf.chef`
* `zookeeper['conf_dir']` - The directory used inside `/etc/zookeeper` and used via the alternatives system. Default `conf.chef`

### Default Attributes

* `hadoop['core_site']['fs.defaultFS']` - Sets URI to HDFS NameNode. Default `hdfs://localhost`
* `hadoop['yarn_site']['yarn.resourcemanager.hostname']` - Sets hostname of YARN ResourceManager. Default `localhost`
* `hive['hive_site']['javax.jdo.option.ConnectionURL']` - Sets JDBC URL. Default `jdbc:derby:;databaseName=/var/lib/hive/metastore/metastore_db;create=true`
* `hive['hive_site']['javax.jdo.option.ConnectionDriverName']` - Sets JDBC Driver. Default `org.apache.derby.jdbc.EmbeddedDriver`

## Recipes

* `default.rb` - Sets up configuration and `hadoop-client` packages.
* `hadoop_hdfs_checkconfig` - Ensures the HDFS configuration meets required parameters.
* `hadoop_hdfs_datanode` - Sets up an HDFS DataNode.
* `hadoop_hdfs_ha_checkconfig` - Ensures the HDFS configuration meets requirements for High Availability.
* `hadoop_hdfs_journalnode` - Sets up an HDFS JournalNode.
* `hadoop_hdfs_namenode` - Sets up an HDFS NameNode.
* `hadoop_hdfs_secondarynamenode` - Sets up an HDFS Secondary NameNode.
* `hadoop_hdfs_zkfc` - Sets up HDFS Failover Controller, required for automated NameNode failover.
* `hadoop_yarn_nodemanager` - Sets up a YARN NodeManager.
* `hadoop_yarn_proxyserver` - Sets up a YARN Web Proxy.
* `hadoop_yarn_resourcemanager` - Sets up a YARN ResourceManager.
* `hbase` - Sets up configuration and `hbase` packages.
* `hbase_checkconfig` - Ensures the HBase configuration meets required parameters.
* `hbase_master` - Sets up an HBase Master.
* `hbase_regionserver` - Sets up an HBase RegionServer.
* `hbase_rest` - Sets up an HBase REST interface.
* `hbase_thrift` - Sets up an HBase Thrift interface.
* `hive` - Sets up configuration and `hive` packages.
* `hive_metastore` - Sets up Hive Metastore metadata repository.
* `hive_server` - Sets up a Hive Thrift service.
* `hive_server2` - Sets up a Hive Thrift service with Kerberos and multi-client concurrency support.
* `oozie` - Sets up an Oozie server.
* `oozie_client` - Sets up an Oozie client.
* `pig` - Installs pig interpreter.
* `repo` - Sets up package manager repositories for specified `hadoop['distribution']`
* `spark` - Sets up configuration and `spark-core` packages.
* `spark_master` - Sets up a Spark Master.
* `spark_worker` - Sets up a Spark Worker.
* `storm` - Sets up `storm` package.
* `storm_nimbus` - Setups a Storm Nimbus server.
* `storm_supervisor` - Setups a Storm Supervisor server.
* `storm_ui` - Setups a Storm UI server.
* `tez` - Sets up configuration  and `tez` packages.
* `zookeeper` - Sets up `zookeeper` package.
* `zookeeper_server` - Sets up a ZooKeeper server.

## Author

Author:: Cask Data, Inc. (<ops@cask.co>)

## Testing

This cookbook has several ways to test it. It includes code tests, which are done using `foodcritic`, `rubocop`, and `chefspec`.
It, also, includes functionality testing, provided by `kitchen`.

```text

CHANGELOG.md
hadoop CHANGELOG
================

v2.14.0 (May 4, 2018)
---------------------
- add TESTING.md ( Issue: #356 )
- add support for HDP 2.6.4.0 ( Issues: #358 #359 )
- add helper to fetch HDP build number from public repository ( Issues: #360 )

v2.13.0 (Dec 5, 2017)
---------------------

- Update HDP GPG key location per docs (Issue: #352)
- Add capability to skip system_tuning recipe (Issue: #353)
- Update testing framework to modern Chef standards (Issue: #354)
- Add support for HDP 2.6.3 (Issue: #355)

v2.12.0 (Sep 12, 2017)
----------------------

- Remove redundant HDP version attributes ( Issue: #344 )
- Update HDP-UTILS repo to 1.1.0.21 ( Issues: #346 [COOK-127](https://issues.cask.co/browse/COOK-127) )
- Use libsnappy1 from hdp-utils on Ubuntu 16 ( Issue: #347 )
- Add Support for HDP 2.6.2.0 and HDP 2.5.6.0 ( Issue: #348 )
- Fix CentOS7 support for HDP, configure only update or GA repo ( Issues: #349 [COOK-129](https://issues.cask.co/browse/COOK-129) )
- Drop support for RHEL 5 and derivatives ( Issues: #351 [COOK-130](https://issues.cask.co/browse/COOK-130) )

v2.11.5 (Aug 7, 2017)
---------------------

- Support sysctl cookbook 0.10.0 on Chef 12.5+ ( Issues: #340 )

v2.11.4 (Aug 7, 2017)
---------------------

- Constrain sysctl due to its Chef version dependency ( Issues: #341 )

v2.11.3 (Jul 28, 2017)
----------------------

- Use correct libsnappy package on Ubuntu 16, per @gwohletz ( Issues: #337 [COOK-126](https://issues.cask.co/browse/COOK-126) )
- Cleanup Gemfile ( Issue: #338 )

v2.11.2 (Jun 20, 2017)
----------------------

- Create hadoop log directory ( Issue: #335 )

v2.11.1 (Jun 19, 2017)
----------------------

- Add HDP 2.5.5.0 ( Issue: #328 )
- Update cookbook headers ( Issue: #329 )
- Support installing Slider ( Issue: #330 )
- Support Hive2 installation and configuration ( Issue: #331 )
- Support Spark 2.x (spark2) package ( Issue: #332 )
- Add HDP 2.6.1 ( Issue: #333 )

v2.11.0 (Apr 18, 2017)
----------------------

- Remove MRv1 daemon support ( Issues: #313 [COOK-118](https://issues.cask.co/browse/COOK-118) )
- Setup NodeManager local directories with defaults ( Issues: #315 [COOK-42](https://issues.cask.co/browse/COOK-42) )
- Create directory for dfs.domain.socket.path ( Issues: #316 [COOK-56](https://issues.cask.co/browse/COOK-56) )
- Create Derby DB directories ( Issues: #317 [COOK-55](https://issues.cask.co/browse/COOK-55) )
- Support `amazon` platform_family for Chef 13 ( Issues: #323 [COOK-123](https://issues.cask.co/browse/COOK-123) )
- Explicitly require 'chef/version' before using Chef::VERSION ( Issue: #324 )
- Properly configure mass_threshold and inline disable FC023 ( Issue: #325 )
- Test Chef 13: remove Ruby 2.0, add Ruby 2.3 ( Issue: #326 )

v2.10.1 (Apr 11, 2017)
----------------------

- Fix issue with HDP 2.6.0.3 ( Issue: #321 )

v2.10.0 (Apr 6, 2017)
---------------------

- Update tests ( Issues: #314 #318 )
- Support HDP 2.6.0.3-8 ( Issue: #319 )

v2.9.0 (Dec 13, 2016)
---------------------

- Update CDH/HDP default versions ( Issue: #310 )
- Support HDP 2.5.3.0 ( Issue: #311 )

v2.8.0 (Dec 6, 2016)
--------------------

- Updates to testing configurations ( Issues: #301 #302 #304 )
- Update HDP versions when short versions given ( Issue: #303 )
- Support 2015 and 2016 Amazon Linux AMIs ( Issue: #305 )
- Use Java cookbook jdk.sh before BigTop tools ( Issues: #306 [COOK-107](https://issues.cask.co/browse/COOK-107) )
- Hive logging support ( Issues: #307 [COOK-111](https://issues.cask.co/browse/COOK-111) )
- Fix hadoop-mapreduce-historyserver stop ( Issue: #308 )

v2.7.1 (Oct 25, 2016)
---------------------

- Switch to new Ruby hash syntax ( Issue #297)
- Add non-cookbook files to chefignore ( Issue #298)
- Support for Spark on IOP ( Issue #299 )

v2.7.0 (Oct 17, 2016)
---------------------

- Pull in code from forks ( Issue #282 )
- Use correct License file ( Issue #283 )
- Set distribution defaults in attributes instead of recipes ( Issue #284 )
- Updated codeclimate configuration ( Issue #285 )
- Updated README ( Issue #286 )
- Support HDP 2.4.3.0, and fix issues with 2.2.1.0, 2.2.6.3, 2.3.6.0 ( Issues #287 #295 )
- Refactoring JAAS configuration, Mapreduce, and Tez ( Issues #288 #293 #292)
- Workaround for a missing dependency in HDP Zookeeper Ubuntu packages ( Issue #289 )
- Update Rubocop configurations ( Issue #290 )
- Create library function for Hadoop properties ( Issue #294 )

v2.6.0 (Sep 9, 2016)
--------------------

- Support for IBM Open Platform ( Issue: #226 )
- Support IOP 4.2.0.0 (and others) ( Issue: #278 )
- HDP: Remove 2.2.1.0 and 2.2.6.3 and add 2.3.6.0 ( Issue: #279 )
- Rename LICENSE to LICENCE.txt ( Issue: #280 )

v2.5.0 (Sep 1, 2016)
--------------------

- Remove `policy_rcd` and use `dpkg_autostart` cookbook ( Issue: #272 )
- Update testing framework ( Issue: #273 )
- Use major_platform_version for centos5 or HDP >= 2.3 ( Issue: #275 )

recipes/repo.rb
#
# Cookbook:: hadoop
# Recipe:: repo
#
# Copyright © 2013-2017 Cask Data, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

major_platform_version = node['platform_version'].to_i
key = 'RPM-GPG-KEY'

# Ensure apt caches are updated and apt resources available
include_recipe 'apt' if node['platform_family'] == 'debian'

case node['hadoop']['distribution']
when 'hdp'
  case node['hadoop']['distribution_version']
  when '2.0.5.0', '2.0.6.0', '2.0.6.1', '2.0.10.0', '2.0.11.0', '2.0.12.0', '2.0.13.0'
    hdp_version = '2.0.4.0'
    hdp_update_version = node['hadoop']['distribution_version']
  when '2.0'
    hdp_version = '2.0.4.0'
    hdp_update_version = '2.0.13.0'
    Chef::Log.warn("Short versions for node['hadoop']['distribution_version'] are deprecated! Please use full version!")
    node.override['hadoop']['distribution_version'] = hdp_update_version
  when '2.0.4.0', '2.1.1.0', '2.2.0.0'
    hdp_version = node['hadoop']['distribution_version']
    hdp_update_version = nil
  when '2.1.2.0', '2.1.2.1', '2.1.3.0', '2.1.4.0', '2.1.5.0', '2.1.7.0', '2.1.10.0', '2.1.15.0'
    hdp_version = '2.1.1.0'
    hdp_update_version = node['hadoop']['distribution_version']
  when '2.1'
    hdp_version = '2.1.1.0'
    hdp_update_version = '2.1.15.0'
    Chef::Log.warn("Short versions for node['hadoop']['distribution_version'] are deprecated! Please use full version!")
    node.override['hadoop']['distribution_version'] = hdp_update_version
  # 2.3, 2.4, and 2.5 do not have their own base version
  when '2.2.1.0', '2.2.4.2', '2.2.4.4', '2.2.6.0', '2.2.6.3', '2.2.8.0', '2.2.9.0',
       '2.3.0.0', '2.3.2.0', '2.3.4.0', '2.3.4.7', '2.3.6.0',
       '2.4.0.0', '2.4.2.0', '2.4.3.0',
       '2.5.0.0', '2.5.3.0', '2.5.5.0', '2.5.6.0',
       '2.6.0.3', '2.6.1.0', '2.6.2.0', '2.6.3.0', '2.6.4.0'
    hdp_version = '2.2.0.0'
    hdp_update_version = node['hadoop']['distribution_version']
  when '2.2'
    hdp_version = '2.2.0.0'
    hdp_update_version = '2.2.9.0'
    Chef::Log.warn("Short versions for node['hadoop']['distribution_version'] are deprecated! Please use full version!")
    node.override['hadoop']['distribution_version'] = hdp_update_version
  when '2.3'
    hdp_version = '2.2.0.0'
    hdp_update_version = '2.3.6.0'
    Chef::Log.warn("Short versions for node['hadoop']['distribution_version'] are deprecated! Please use full version!")
    node.override['hadoop']['distribution_version'] = hdp_update_version
  when '2.4'
    hdp_version = '2.2.0.0'
    hdp_update_version = '2.4.3.0'
    Chef::Log.warn("Short versions for node['hadoop']['distribution_version'] are deprecated! Please use full version!")
    node.override['hadoop']['distribution_version'] = hdp_update_version
  when '2.5'
    hdp_version = '2.2.0.0'
    hdp_update_version = '2.5.6.0'
    Chef::Log.warn("Short versions for node['hadoop']['distribution_version'] are deprecated! Please use full version!")
    node.override['hadoop']['distribution_version'] = hdp_update_version
  when '2.6', '2'
    hdp_version = '2.2.0.0'
    hdp_update_version = '2.6.4.0'
    Chef::Log.warn("Short versions for node['hadoop']['distribution_version'] are deprecated! Please use full version!")
    node.override['hadoop']['distribution_version'] = hdp_update_version
  else
    Chef::Application.fatal!('This cookbook only supports HDP 2.x')
  end

  hdp_utils_version = '1.1.0.21'

  case node['platform_family']
  when 'rhel', 'amazon'
    yum_base_url = 'http://public-repo-1.hortonworks.com/HDP'
    os = if hdp_update_version.to_f >= 2.3
           "centos#{major_platform_version}"
         else
           'centos6'
         end

    if hdp_update_version.nil?
      # We are on one of the GA versions; configure the GA repo
      # rubocop:disable Metrics/BlockNesting
      yum_repo_url = node['hadoop']['yum_repo_url'] ? node['hadoop']['yum_repo_url'] : "#{yum_base_url}/#{os}/2.x/GA/#{hdp_version}"
      yum_repo_key_url = node['hadoop']['yum_repo_key_url'] ? node['hadoop']['yum_repo_key_url'] : "#{yum_repo_url}/#{key}/#{key}-Jenkins"
      # rubocop:enable Metrics/BlockNesting

      yum_repository 'hdp' do
        name 'HDP-2.x'
        description 'Hortonworks Data Platform Version - HDP-2.x'
        url yum_repo_url
        gpgkey yum_repo_key_url
        action :add
      end
    else
      # We are on an update version; configure the update repo only
      # rubocop:disable Metrics/BlockNesting
      yum_repo_url = node['hadoop']['yum_repo_url'] ? node['hadoop']['yum_repo_url'] : "#{yum_base_url}/#{os}/2.x/updates/#{hdp_update_version}"
      yum_repo_key_url = node['hadoop']['yum_repo_key_url'] ? node['hadoop']['yum_repo_key_url'] : "#{yum_repo_url}/#{key}/#{key}-Jenkins"
      # rubocop:enable Metrics/BlockNesting

      yum_repository 'hdp-updates' do
        name 'Updates-HDP-2.x'
        description 'Updates for Hortonworks Data Platform Version - HDP-2.x'
        url yum_repo_url
        gpgkey yum_repo_key_url
        action :add
      end

      # HDP 2.6.4.0 has moved hadooplzo to a separate repo
      yum_repository 'hdp-updates' do
        name 'HDP-GPL-2.x'
        description 'Hortonworks Data Platform Version - HDP-GPL-2.x'
        url "#{yum_base_url}-GPL/#{os}/2.x/updates/#{hdp_update_version}"
        gpgkey yum_repo_key_url
        action :add
        only_if { Gem::Version.new(hdp_update_version) >= Gem::Version.new('2.6.4.0') }
      end

    end
    yum_repository 'hdp-utils' do
      name "HDP-UTILS-#{hdp_utils_version}"
      description "Hortonworks Data Platform Utils Version - HDP-UTILS-#{hdp_utils_version}"
      url "#{yum_base_url}-UTILS-#{hdp_utils_version}/repos/#{os}"
      gpgkey yum_repo_key_url
      action :add
    end

  when 'debian'
    apt_domain_name = 'public-repo-1.hortonworks.com'
    apt_base_url = "http://#{apt_domain_name}/HDP"
    hdp_update_version = hdp_version if hdp_update_version.nil?
    # HDP supports Debian 6, Ubuntu 12. Starting with 2.3.2.0 Ubuntu 14 and Debian 7
    if node['platform'] == 'debian' && node['platform_version'] =~ /^7/ && Gem::Version.new(hdp_update_version) >= Gem::Version.new('2.3.2.0')
      os = "#{node['platform']}7"
    elsif node['platform'] == 'debian'
      os = "#{node['platform']}6"
    elsif node['platform'] == 'ubuntu' && node['platform_version'] =~ /^14/ && Gem::Version.new(hdp_update_version) >= Gem::Version.new('2.3.2.0')
      os = "#{node['platform']}14"
    elsif node['platform'] == 'ubuntu'
      os = "#{node['platform']}12"
    end
    hdp_apt_repo_path =
      case hdp_update_version
      when '2.2.0.0'
        "2.x/GA/#{hdp_update_version}"
      when '2.1.10.0', '2.1.15.0', '2.2.1.0', '2.2.4.2', '2.2.6.0', '2.2.6.3', '2.2.8.0', '2.2.9.0',
           '2.3.0.0', '2.3.2.0', '2.3.4.0', '2.3.4.7', '2.3.6.0',
           '2.4.0.0', '2.4.2.0', '2.4.3.0',
           '2.5.0.0', '2.5.3.0', '2.5.5.0', '2.5.6.0',
           '2.6.0.3', '2.6.1.0', '2.6.2.0', '2.6.3.0', '2.6.4.0'
        "2.x/updates/#{hdp_update_version}"
      else
        hdp_update_version
      end
    apt_repo_url = node['hadoop']['apt_repo_url'] ? node['hadoop']['apt_repo_url'] : "#{apt_base_url}/#{os}/#{hdp_apt_repo_path}"
    # Hortonworks don't know how to provide a key, but we do
    apt_repo_key_url = node['hadoop']['apt_repo_key_url'] ? node['hadoop']['apt_repo_key_url'] : "#{apt_base_url}/centos6/#{key}/#{key}-Jenkins"

    apt_repository 'hdp' do
      uri apt_repo_url
      key apt_repo_key_url
      distribution 'HDP'
      trusted true
      components ['main']
      action :add
    end

    # HDP 2.6.4.0 has moved hadooplzo to a separate repo
    apt_repository 'hdp-gpl' do
      uri "#{apt_base_url}-GPL/#{os}/2.x/updates/#{hdp_update_version}"
      key apt_repo_key_url
      distribution 'HDP-GPL'
      trusted true
      components ['main']
      action :add
      only_if { Gem::Version.new(hdp_update_version) >= Gem::Version.new('2.6.4.0') }
    end

    apt_repository 'hdp-utils' do
      uri "#{apt_base_url}-UTILS-#{hdp_utils_version}/repos/#{os}"
      key apt_repo_key_url
      distribution 'HDP-UTILS'
      components ['main']
      action :add
    end
    apt_preference 'hdp' do
      glob '*'
      pin "origin #{apt_domain_name}"
      pin_priority '700'
    end
  end # End hdp

when 'cdh'
  cdh_release = node['hadoop']['distribution_version'].to_i
  if node['hadoop']['distribution_version'].to_f >= 5.3 && node.key?('java') && node['java'].key?('jdk_version') && node['java']['jdk_version'].to_i < 7
    Chef::Application.fatal!('CDH 5.3 and above require Java 7 or higher')
  end
  Chef::Log.warn("Short versions for node['hadoop']['distribution_version'] are deprecated! Please use full version!") if node['hadoop']['distribution_version'].to_s == '5'
  case node['platform_family']
  when 'rhel', 'amazon'
    yum_base_url = "http://archive.cloudera.com/cdh#{cdh_release}/redhat"
    yum_repo_url = node['hadoop']['yum_repo_url'] ? node['hadoop']['yum_repo_url'] : "#{yum_base_url}/#{major_platform_version}/#{node['kernel']['machine']}/cdh/#{node['hadoop']['distribution_version']}"
    yum_repo_key_url = node['hadoop']['yum_repo_key_url'] ? node['hadoop']['yum_repo_key_url'] : "#{yum_base_url}/#{major_platform_version}/#{node['kernel']['machine']}/cdh/#{key}-cloudera"

    yum_repository "cloudera-cdh#{cdh_release}" do
      name "cloudera-cdh#{cdh_release}"
      description "Cloudera's Distribution for Hadoop, Version #{cdh_release}"
      url yum_repo_url
      gpgkey yum_repo_key_url
      action :add
    end

  when 'debian'
    codename = node['lsb']['codename']

    # rubocop: disable Metrics/BlockNesting
    case codename
    when 'raring', 'saucy'
      Chef::Log.warn('This version of Ubuntu is unsupported by Cloudera! Bug reports should include patches.')
      codename = 'precise'
    when 'trusty'
      unless cdh_release >= 5
        Chef::Log.warn('This version of Ubuntu is unsupported by Cloudera! Bug reports should include patches.')
        codename = 'precise'
      end
    end
    # rubocop: enable Metrics/BlockNesting

    apt_domain_name = 'archive.cloudera.com'
    apt_base_url = "http://#{apt_domain_name}/cdh#{cdh_release}/#{node['platform']}"
    apt_repo_url = node['hadoop']['apt_repo_url'] ? node['hadoop']['apt_repo_url'] : "#{apt_base_url}/#{codename}/amd64/cdh"
    apt_repo_key_url = node['hadoop']['apt_repo_key_url'] ? node['hadoop']['apt_repo_key_url'] : "#{apt_base_url}/#{codename}/amd64/cdh/archive.key"

    apt_repository "cloudera-cdh#{cdh_release}" do
      uri apt_repo_url
      key apt_repo_key_url
      distribution "#{codename}-cdh#{node['hadoop']['distribution_version']}"
      components ['contrib']
      arch 'amd64'
      action :add
    end

    apt_preference "cloudera-cdh#{cdh_release}" do
      glob '*'
      pin "origin #{apt_domain_name}"
      pin_priority '700'
    end
  end # End cdh

when 'bigtop'
  bigtop_release = node['hadoop']['distribution_version']

  # allow a developer mode for use when developing against bigtop, see https://issues.cask.co/browse/COOK-1
  if bigtop_release.casecmp('develop') == 0 && !(node['hadoop'].key?('yum_repo_url') || node['hadoop'].key?('apt_repo_url'))
    Chef::Application.fatal!("You must set node['hadoop']['yum_repo_url'] or node['hadoop']['apt_repo_url'] when specifying node['hadoop']['distribution_version'] == 'develop'")
  end

  # do not validate gpg repo keys when in develop mode
  validate_repo_key = bigtop_release.casecmp('develop') == 0 ? false : true
  Chef::Log.warn('Allowing install of unsigned binaries') unless validate_repo_key

  case node['platform_family']
  when 'rhel', 'amazon'

    case major_platform_version
    when 5, 6
      yum_platform_version = major_platform_version
    when 2014, 2015, 2016 # Amazon Linux, point to redhat/6 bigtop repo
      yum_platform_version = 6
    else
      Chef::Log.warn('Unsupported platform detected, use at your own risk')
      yum_platform_version = major_platform_version
    end

    yum_base_url = "http://bigtop-repos.s3.amazonaws.com/releases/#{bigtop_release}/centos"
    yum_repo_url = node['hadoop']['yum_repo_url'] ? node['hadoop']['yum_repo_url'] : "#{yum_base_url}/#{yum_platform_version}/#{node['kernel']['machine']}"
    yum_repo_key_url = node['hadoop']['yum_repo_key_url'] ? node['hadoop']['yum_repo_key_url'] : 'http://archive.apache.org/dist/bigtop/KEYS'

    yum_repository "bigtop-#{bigtop_release}" do
      name "bigtop-#{bigtop_release}"
      description "Apache Bigtop Distribution for Hadoop, Version #{bigtop_release}"
      url yum_repo_url
      gpgkey yum_repo_key_url
      gpgcheck validate_repo_key
      action :add
    end

  when 'debian'
    # for bigtop, we do not validate codename, to support developing against custom repositories
    codename = node['lsb']['codename']

    # rubocop: disable Metrics/BlockNesting
    case codename
    when 'precise', 'quantal', 'raring', 'saucy', 'utopic', 'vivid', 'wily'
      if bigtop_release.to_f >= 1.0
        Chef::Log.warn('This version of Ubuntu is unsupported by Bigtop! Bug reports should include patches.')
        codename = 'trusty'
      end
    end
    # rubocop: enable Metrics/BlockNesting

    apt_domain_name = 'bigtop-repos.s3.amazonaws.com'
    apt_base_url = "http://#{apt_domain_name}/releases/#{bigtop_release}/#{node['platform']}"
    apt_repo_url = node['hadoop']['apt_repo_url'] ? node['hadoop']['apt_repo_url'] : "#{apt_base_url}/#{codename}/#{node['kernel']['machine']}"
    apt_repo_key_url = node['hadoop']['apt_repo_key_url'] ? node['hadoop']['apt_repo_key_url'] : 'http://archive.apache.org/dist/bigtop/KEYS'

    apt_repository "bigtop-#{bigtop_release}" do
      uri apt_repo_url
      key apt_repo_key_url
      trusted !validate_repo_key
      distribution 'bigtop'
      components ['contrib']
      action :add
    end
    apt_preference "bigtop-#{bigtop_release}" do
      glob '*'
      pin "origin #{apt_domain_name}"
      pin_priority '700'
    end
  end

when 'iop'
  iop_version = node['hadoop']['distribution_version']
  iop_release = "#{node['hadoop']['distribution_version'].to_f}.x"
  iop_utils_version = '1.2.0.0'

  case node['platform_family']
  when 'rhel', 'amazon'
    # https://ibm-open-platform.ibm.com/repos/IOP/rhel/6/x86_64/4.1.x/GA/4.1.0.0/
    yum_base_url = 'https://ibm-open-platform.ibm.com/repos/IOP'
    os = 'rhel'
    v = major_platform_version
    m = node['kernel']['machine']
    key = 'BI-GPG-KEY.public'

    yum_repo_url = node['hadoop']['yum_repo_url'] ? node['hadoop']['yum_repo_url'] : "#{yum_base_url}/#{os}/#{v}/#{m}/#{iop_release}/GA/#{iop_version}"
    yum_repo_key_url = node['hadoop']['yum_repo_key_url'] ? node['hadoop']['yum_repo_key_url'] : "#{yum_repo_url}/#{key}"

    yum_repository 'iop' do
      name "IOP-#{iop_release}"
      description "IBM Open Platform Version - IOP-#{iop_release}"
      url yum_repo_url
      gpgkey yum_repo_key_url
      action :add
    end
    yum_repository 'iop-utils' do
      name "IOP-UTILS-#{iop_utils_version}"
      description "IBM Open Platform Utils Version - IOP-UTILS-#{iop_utils_version}"
      url "#{yum_base_url}-UTILS/#{os}/#{v}/#{m}/#{iop_utils_version.to_f}"
      gpgkey yum_repo_key_url
      # disable gpgcheck because the Hadoop-lzo package is unsigned
      gpgcheck false
      action :add
    end
  else
    Chef::Application.fatal!("IBM Open Platform only supports RHEL-family! You're on #{node['platform_family']}!")
  end

else
  # COOK-25 fail fast
  Chef::Application.fatal!("Invalid node['hadoop']['distribution'] (#{node['hadoop']['distribution']}) specified!")
end
